### DISCUSSION TITLE
Need help trying to define a custom surround (deletion)

### DISCUSSION THREAD

**Discussion by smjonas** [kylechui/nvim-surround#121] [Q&A] [answered]
I am trying to create my own custom surrounds but I can't get this one to work. @kylechui Any idea why the following doesn't do anything when I run `ds*` (e.g. when the line is `**word**` and the cursor is on `w`)?  

```lua
require("nvim-surround").buffer_setup {
  surrounds = {
    ["*"] = {
      find = function()
        return require("nvim-surround.config").get_selection { textobject = "*" }
      end,
      delete = "^(%*%*?)().-(%*%*?)()$",
    },
  },
}
```
Also, is the `find` key even required? I couldn't find anything in the docs and haven't looked at the code yet. If that is currently the case, wouldn't it be possible to infer the selections from the deletion pattern? Otherwise a warning might be good to let users know that their surround won't work :)

**Comment by kylechui** [comment, marked as answer]
Here you're going to want to use `find = "%*%*?.-%*%*?"`. This is because the text-object `a*` *does not* exist (unless it is defined by another plugin, e.g. `targets.vim`). The idea is that the `textobject` key should only really be used for built-in text-objects or for characters for which there exists a defined text-object, e.g. `T` key uses `t` for finding the parent selection. I'll look into improving the docs; the `find` key *can* be omitted but it defaults to `invalid_key_behavior`'s `find` key, which just looks for a surrounding pair of the character you inputted.

In general, the process is:
* The `find` key finds a "parent" selection that includes the surrounding pair
  * If omitted, `nvim-surround` will fallback on what's defined by `invalid_key_behavior`
  * The `textobject` key finds the "nearest" Vim text-object and uses that as the selection
  * The `pattern` key uses a Lua pattern to match the parent selection
* The `delete`/`change.target` keys will then narrow the parent selection down to the pair of selections you would like to delete/change

General rule of thumb: If you can't do `va[char]` and highlight a selection, then the text-object `a[char]` doesn't exist

Also I assume you mean `ds*` in your question :)

**Reply by smjonas** [reply]
Thank you for the explanation, that worked! 

> This is because the text-object a* does not exist  (unless it is defined by another plugin, e.g. targets.vim)

It is indeed defined by `targets.vim`, I did not even consider that it's not a standard text object but you're right :D 

> Also I assume you mean ds* in your question :)

Yes, I updated the question :)

**Reply by kylechui** [reply]
Does `va*` select around a pair of asterisks like I think it should? If that's the case, then theoretically what you originally had should work as well :thinking: I'll re-install `targets.vim` later today and give it a go

**Reply by kylechui** [reply]
Reading through the documentation, it seems like targets treats those "separator" characters much like `"`, where `a*` doesn't actually select *both* of the characters. In this case, it seems like `A*` would make it work, but I didn't take into account that people might want to use *other* (non `a[char]`) text-objects :/ Try using my suggestion above, but I would need a minor breaking change in order to give a "proper" compatibility fix with targets.

**Reply by smjonas** [reply]
For some reason `va*` won't include the last asterisk in the selection, maybe that's why it couldn't find the pair of `*`. That's seems to be strange behavior from `targets.vim`.

Edit: beat me to it in your edit :smile:

**Reply by kylechui** [reply]
Also another thing I noticed is that `vA*` doesn't actually select multiple pairs?, e.g. for `**hello**` it would only select the inner pair, so you couldn't delete double asterisks with it. However, if you're okay with that and don't mind a bit of *jank*, here's what you can do to utilize it:
```lua
require("nvim-surround").setup({
    surrounds = {
        ["*"] = {
            find = function()
                local esc = vim.api.nvim_replace_termcodes("<esc>", true, false, true)
                return require("nvim-surround.config").get_selection({ textobject = esc .. "g@A*" })
            end,
            delete = "^(%*%*?)().-(%*%*?)()$",
        },
    },
})
```
The text-object key actually just runs a normal command with `"normal g@a" .. char` behind the scenes, so you can just "cancel" it and run the desired one.

**Reply by kylechui** [reply]
Alternatively you can use the following:
```lua
require("nvim-surround").setup({
    surrounds = {
        ["*"] = {
            find = "(%*+).-(%*+)",
            delete = "^(%*+)().-(%*+)()$",
        },
    },
})
```
which will delete all sets of asterisk pairs. The only caveat is that imbalanced pairs will not be deleted. However, if that's desired I'm sure there's a way to do it using the given `config.get_selection` and `config.get_selections` functions, where you validate matched pairs before returning.
